package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -0,0 +1,351 @@
+        ----------
+        filename : str or :class:`h5py.File`
+            trajectory filename or open h5py file
+        n_atoms : int
+            number of atoms in trajectory
+        convert_units : bool (optional)
+            convert units to MDAnalysis units
+        **kwargs : dict
+
+        """
+        self.filename = filename
+        if n_atoms == 0:
+            raise ValueError("H5MDWriter: no atoms in output trajectory")
+        self.n_atoms = n_atoms
+        self.convert_units = convert_units
+
+        self.h5mdfile = None
+        self.has_positions = kwargs.get('positions', False)
+        self.has_velocities = kwargs.get('velocities', False)
+        self.has_forces = kwargs.get('forces', False)
+
+        # Pull out various keywords to store in 'h5md' group
+        self.author = kwargs.pop('author', 'N/A')
+        self.author_email = kwargs.pop('author_email', None)
+        self.creator = kwargs.pop('creator', 'MDAnalysis')
+        self.creator_version = kwargs.pop('creator_version', mda.__version__)
+
+        self.units = units
+
+    def _init_h5md(self, periodic=True):
+        """Initializes H5MD trajectory.
+
+        The `H5MD`_ file is opened using the `H5PY`_ library. """
+
+        h5md_file = h5py.File(self.filename, 'w')
+        h5md_group = h5md_file.create_group('h5md')
+        h5md_group.attrs['version'] = np.array([1,1])
+        author_group = h5md_file.create_group('h5md/author')
+        author_group.attrs['name'] = self.author
+        if self.author_email is not None:
+            author_group.attrs['email'] = self.author_email
+        creator_group = h5md_file.create_group('h5md/creator')
+        creator_group.attrs['name'] = self.creator
+        creator_group.attrs['version'] = self.creator_version
+
+        self.h5mdfile = h5md_file
+        self._trajectory_group = h5mdfile.require_group('particles').require_group('trajectory')
+
+    def is_periodic(self, ts):
+        """Test if timestep ``ts`` contains a periodic box.
+
+        Parameters
+        ----------
+        ts : :class:`Timestep`
+             :class:`Timestep` instance containing coordinates to
+             be written to trajectory file
+
+        Returns
+        -------
+        bool
+            Return ``True`` if `ts` contains a valid simulation box
+        """
+        return np.all(ts.dimensions > 0)
+
+    def _write_next_frame(self, ag):
+        """Write information associated with ``ag`` at current frame
+        into trajectory
+
+        Parameters
+        ----------
+        ag : AtomGroup or Universe
+
+        .. versionadded:: 2.0.0
+
+        """
+        try:
+            # Atomgroup?
+            ts = ag.ts
+        except AttributeError:
+            try:
+                # Universe?
+                ts = ag.trajectory.ts
+            except AttributeError:
+                errmsg = "Input obj is neither an AtomGroup or Universe"
+                raise TypeError(errmsg) from None
```

Comment:
For this block I'm also reminded of cases in the code base that look something like this:

```python
        if isinstance(arg, Universe):
            # action 1
        elif isinstance(arg, AtomGroup):
            # action 2
        else:
            raise ValueError("arg is not a Universe or AtomGroup")
```

I don't know if it is formally cleander than the try/except, maybe it is just the nesting that made me think an alternative might be desired.

Reply:
Maybe 

```python
try:
    ts = ag.universe.trajectory.ts
except AttributeError:
    ...
```
?

I think there's a performance difference to try/except vs if/else, where try/except is faster unless the exception is raised.

Reply:
Python Zen says "It's easier to ask for forgiveness than to ask for permission." (which supposedly means that idiomatic Python uses try/except instead of if when possible). The try/except also allows duck-typing (if so desired).

For the performance question: could be tested.

Reply:
I think it is just the nesting that caught my attention really--if the approach from @lilyminium is equivalent that may be nicer.

If you're really keen on avoiding the conditionals, looks like there are some opportunities in the code base.

Reply:
I doubt it is worth my energy though.